apps/api/src/auth/auth.service.ts

```diff
@@ -0,0 +1,131 @@
+import {
+    ConflictException,
+    Injectable,
+    InternalServerErrorException,
+    NotFoundException,
+    UnauthorizedException,
+} from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { hashSync } from 'bcryptjs';
+import { randomBytes } from 'crypto';
+import { Repository } from 'typeorm';
+
+import { AccountStatus, Token, TokenType, User } from '@ecoside/database';
+import {
+    ActivateAccountDto,
+    RegisterDto,
+    RequestPasswordResetDto,
+    ResetPasswordDto,
+} from '@ecoside/dto';
+import { EmailService } from '../email';
+import { UserService } from '../user';
+
+@Injectable()
+export class AuthService {
+    constructor(
+        @InjectRepository(User) private users: Repository<User>,
+        @InjectRepository(Token) private tokens: Repository<Token>,
+        private userService: UserService,
+        private emailService: EmailService,
+    ) {}
+
+    async login(email: string, password: string) {
+        try {
+            const user = await this.userService.findOne(email);
+
+            if (user.isAccountActive() && user.verifyPassword(password)) {
+                return user;
+            } else {
+                throw new UnauthorizedException();
+            }
+        } catch {
+            throw new UnauthorizedException();
+        }
+    }
+
+    async register(dto: RegisterDto) {
+        const user = this.users.create(dto);
+        user.password_hash = hashSync(dto.password);
+
+        try {
+            await this.users.save(user);
+        } catch {
+            throw new ConflictException(
```

Comment:
Is it only error that can occur here?

Reply:
No there are many more errors possible but checking for error codes everytime you save something is a hassle to do. I thought about it and maybe there is a way to make our own Repository class that would handle all the other errors.

Reply:
Yeeeeah, but 99% of the times the error is regarding the unique constraint on email column. Other cases that .save() throws an error is when there is a issue with schema. In production it is basically the only error that would occur here.
It is good enough but I will try to implement some kind of custom repository / error handler.